index.html

```diff
@@ -819,84 +860,79 @@ <h3><dfn>RTCQuicTransportState</dfn> Enum</h3>
+    <h2><dfn>QUIC Stream API</dfn></h2>
+    <p>The <code>QUIC Stream API</code> includes information relating
     to a QUIC stream. </p>
     <section id="rtcquicstream-overview*">
       <h3>Overview</h3>
-      <p>An <code><a>RTCQuicStream</a></code> instance is associated to
-      an <code><a>RTCQuicTransport</a></code> instance.</p>
+      <p><code><a>RTCQuicBidirectionalStream</a></code>, <code><a>RTCQuicSendStream</a></code>,
+      and <code><a>RTCQuicReceiveStream</a></code> instances are associated to
+      a <code><a>RTCQuicTransport</a></code> instance.</p>
     </section>
     <section id="rtcquicstream-operation*">
       <h3>Operation</h3>
-      <p>An <code><a>RTCQuicStream</a></code> instance is constructed
-      using the <code><a>RTCQuicTransport</a>.createStream</code> method.</p>
+      <p>A <code><a>RTCQuicBidirectionalStream</a></code> can be created in the
+      following ways:</p>
+      <ol>
+        <li>Using the <code><a>RTCQuicTransport</a>.createBidirectionalStream</code></li>
+        <li>Getting a <code><a>newbidirectionalstreamwithdata</a></code> event on the
+        <code><a>RTCQuicTransport</a></code></li>
+      </ol>
+      <p>A <code><a>RTCQuicSendStream</a></code> can be created in the following
+      ways:</p>
+      <ol>
+        <li>Using the <code><a>RTCQuicTransport</a>.createSendStream</code></li>
+      </ol>
+      <p>A <code><a>RTCQuicReceiveStream</a></code> can be created in the following
+      ways:</p>
+      <ol>
+        <li>Getting a <code><a>newreceivestreamwithdata</a></code> event on the
+        <code><a>RTCQuicTransport</a></code></li>
+      </ol>
     </section>
-    <section id="rtcquicstream-interface-definition*">
-      <h3>Interface Definition</h3>
+    <section id="rtcquicwritablestream-interface-mixin-definition*">
+      <h3>Interface Mixin <dfn>RTCQuicWritableStream</dfn></h3>
       <div>
         <pre class="idl">
         [ Exposed=Window ]
-interface RTCQuicStream {
-    readonly        attribute RTCQuicTransport    transport;
-    readonly        attribute RTCQuicStreamState  state;
-    readonly        attribute unsigned long       readableAmount;
-    readonly        attribute unsigned long       writeBufferedAmount;
-    RTCQuicStreamParameters getParameters();
-    long                  readInto (Uint8Array data);
-    void                  write (Uint8Array data);
-    void                  finish ();
-    void                  reset ();
-    Promise&lt;void&gt;   waitForReadable(unsigned long amount);
-    Promise&lt;void&gt;   waitForWriteBufferedAmountBelow(unsigned long threshold);
-                    attribute EventHandler        onstatechange;
-};</pre>
+        interface mixin RTCQuicWritableStream {
+            readonly attribute boolean writable;
+            readonly attribute unsigned long writeBufferedAmount;
+            readonly attribute Promise&lt;RTCQuicStreamAbortInfo&gt; writingAborted;

```

Comment:
I think it's nice for the reasons you mentioned:
- You can get it after it has fired. If the promise has already been settled, it will still fulfill/reject appropriately.
- Making it an attribute means that there should only be one promise that will only settled once.
- We should only get a callback for a RST_STREAM/STOP_SENDING frame once

Here's a blog post where I learned about promises and explains a tiny amount about when they're more appropriate than events (spoiler alert it's pretty much what you said).
https://developers.google.com/web/fundamentals/primers/promises#events_arent_always_the_best_way

Reply:
I've never seen it used either, but it is mentioned in WebIDL:
https://heycam.github.io/webidl/#idl-promise

Reply:
I don't disagree that an event could work here, but a promise seems more appropriate. The WHATWG streams do the same thing:
https://streams.spec.whatwg.org/#default-reader-closed
https://streams.spec.whatwg.org/#default-writer-closed

Reply:
I agree that `Promise`s are a good alternative to *fire once* events.

Reply:
How about I keep this as a Promise in this PR, and we can move this discussion for after it merges if anyone still feels strongly for using an event here?

Reply:
Sounds ok.